src/rtc_signaling.cpp

```diff
@@ -0,0 +1,430 @@
+        if (!self) return;
+
+		post(m_io_context, std::bind(&rtc_signaling::on_data_channel
+			, self
+			, error_code{}
+			, offer_id
+			, dc
+        ));
+    });
+
+	time_duration const timeout = seconds(m_torrent->settings().get_int(settings_pack::webtorrent_connection_timeout));
+	connection conn(m_io_context);
+	conn.peer_connection = std::move(pc);
+	conn.timer.expires_after(timeout);
+	conn.timer.async_wait(std::bind(&rtc_signaling::on_data_channel
+		, shared_from_this()
+        , boost::asio::error::timed_out
+		, offer_id
+        , nullptr
+    ));
+
+	auto it = m_connections.emplace(offer_id, std::move(conn)).first;
+	return it->second;
+}
+
+void rtc_signaling::on_generated_offer(error_code const& ec, rtc_offer offer)
+{
+#ifndef TORRENT_DISABLE_LOGGING
+	debug_log("*** RTC signaling generated offer");
+#endif
+	while (!m_offer_batches.empty() && m_offer_batches.front().is_complete())
+		m_offer_batches.pop();
+
+	if (!m_offer_batches.empty())
+		m_offer_batches.front().add(ec, std::forward<rtc_offer>(offer));
+}
+
+void rtc_signaling::on_generated_answer(error_code const& ec, rtc_answer answer, rtc_offer offer)
+{
+    if (ec)
+    {
+        // Ignore
+        return;
+    }
+#ifndef TORRENT_DISABLE_LOGGING
+	debug_log("*** RTC signaling generated answer");
+#endif
+	TORRENT_ASSERT(offer.answer_callback);
+	peer_id pid = aux::generate_peer_id(m_torrent->settings());
+	offer.answer_callback(pid, answer);
+}
+
+void rtc_signaling::on_data_channel(error_code const& ec
+		, rtc_offer_id offer_id
+		, std::shared_ptr<rtc::DataChannel> dc)
+{
+	auto it = m_connections.find(offer_id);
+	if (it == m_connections.end()) return;
+
+	auto conn = std::move(it->second);
+	m_connections.erase(it);
+
+	if (ec || !conn.pid)
+	{
+#ifndef TORRENT_DISABLE_LOGGING
+		debug_log("*** RTC negotiation failed");
+#endif
+		return;
+	}
+
+#ifndef TORRENT_DISABLE_LOGGING
+	debug_log("*** RTC data channel open");
+#endif
+
+	TORRENT_ASSERT(dc);
+	m_rtc_stream_handler(*conn.pid, rtc_stream_init{conn.peer_connection, dc});
+}
+
+rtc_signaling::offer_batch::offer_batch(int count, rtc_signaling::offers_handler handler)
+	: m_count(count)
+	, m_handler(std::move(handler))
+{
+	if (m_count == 0) m_handler(error_code{}, {});
+}
+
+void rtc_signaling::offer_batch::add(error_code const& ec, rtc_offer offer)
+{
+	if (!ec) m_offers.push_back(std::move(offer));
+	else --m_count;
+
+	if (is_complete()) m_handler(error_code{}, m_offers);
+}
+
+bool rtc_signaling::offer_batch::is_complete() const
+{
+	return int(m_offers.size()) == m_count;
+}
+
+#ifndef TORRENT_DISABLE_LOGGING
+bool rtc_signaling::should_log() const
+{
+	return alerts().should_post<torrent_log_alert>();
+}
+
+TORRENT_FORMAT(2,3)
+void rtc_signaling::debug_log(char const* fmt, ...) const noexcept try
+{
+	if (!alerts().should_post<torrent_log_alert>()) return;
+
+	va_list v;
+	va_start(v, fmt);
+	alerts().emplace_alert<torrent_log_alert>(const_cast<torrent*>(m_torrent)->get_handle(), fmt, v);
```

Comment:
I don't see `debug_log()` being called inside a `const` member function. Are you sure it needs to be `const` itself? If not, this `const_cast<>` could be removed.

Btw, this is not run on a separate thread, right?

Reply:
No, it is called only from posted tasks.